cpp/net/url_fetcher_test.cc

```diff
@@ -0,0 +1,162 @@
+#include <sys/wait.h>
+#include <unistd.h>
+
+#include "net/connection_pool.h"
+#include "net/url_fetcher.h"
+#include "util/libevent_wrapper.h"
+#include "util/status_test_util.h"
+#include "util/sync_task.h"
+#include "util/testing.h"
+#include "util/thread_pool.h"
+
+DECLARE_string(trusted_root_certs);
+
+namespace cert_trans {
+
+using std::make_shared;
+using std::shared_ptr;
+using std::string;
+using std::to_string;
+using util::SyncTask;
+using util::testing::StatusIs;
+
+DEFINE_string(cert_dir, "test/testdata/urlfetcher_test_certs",
+              "Directory containing the test certs.");
+
+
+const uint16_t kLocalHostPort = 4433;
+const uint16_t kNonLocalHostPort = 4434;
+
+
+namespace {
+
+
+pid_t RunOpenSSLServer(uint16_t port, const std::string& cert_file,
+                       const std::string& key_file) {
+  pid_t pid(fork());
+  if (pid == -1) {
+    LOG(INFO) << "fork() failed: " << pid;
+  } else if (pid == 0) {
+    const string port_str(to_string(port));
+    const string cert_str(FLAGS_cert_dir + "/" + cert_file);
+    const string key_str(FLAGS_cert_dir + "/" + key_file);
+    const char* argv[]{"openssl", "s_server",
+                       "-accept", port_str.c_str(),
+                       "-cert",   cert_str.c_str(),
+                       "-key",    key_str.c_str(),
+                       "-www",    0L};
+    execvp(argv[0], const_cast<char**>(argv));
+  }
+  return pid;
+}
+
+
+}  // namespace
+
+
+class UrlFetcherTest : public ::testing::Test {
+ public:
+  UrlFetcherTest()
+      : base_(make_shared<libevent::Base>()), event_pump_(base_), pool_(1) {
+    FLAGS_trusted_root_certs = FLAGS_cert_dir + "/ca-cert.pem";
+    fetcher_.reset(new UrlFetcher(base_.get()));
+  }
+
+ protected:
+  shared_ptr<libevent::Base> base_;
+  shared_ptr<UrlFetcher> fetcher_;
+  libevent::EventPumpThread event_pump_;
+  ThreadPool pool_;
+};
+
+
+TEST_F(UrlFetcherTest, TestCertMatchesHost) {
+  UrlFetcher::Request req(
+      URL("https://localhost:" + to_string(kLocalHostPort)));
+  UrlFetcher::Response resp;
+
+  SyncTask task(&pool_);
+  fetcher_->Fetch(req, &resp, task.task());
+  task.Wait();
+  EXPECT_EQ(util::Status::OK, task.status());
+  EXPECT_EQ(200, resp.status_code);
+}
+
+
+TEST_F(UrlFetcherTest, TestCertDoesNotMatchHost) {
+  UrlFetcher::Request req(
+      URL("https://localhost:" + to_string(kNonLocalHostPort)));
+  UrlFetcher::Response resp;
+
+  SyncTask task(&pool_);
+  fetcher_->Fetch(req, &resp, task.task());
+  task.Wait();
+  EXPECT_THAT(task.status(), StatusIs(util::error::FAILED_PRECONDITION));
+  EXPECT_EQ(kSSLErrorStatus, resp.status_code);
+}
+
+
+TEST_F(UrlFetcherTest, TestNotListening) {
+  UrlFetcher::Request req(URL("https://localhost:63544"));
+  UrlFetcher::Response resp;
+
+  SyncTask task(&pool_);
+  fetcher_->Fetch(req, &resp, task.task());
+  task.Wait();
+  EXPECT_THAT(task.status(), StatusIs(util::error::FAILED_PRECONDITION));
+  EXPECT_EQ(kUnknownErrorStatus, resp.status_code);
+}
```

Comment:
On 12 June 2015 at 11:42, Al Cutter <notifications@github.com> wrote:

> In cpp/net/url_fetcher_test.cc
> <https://github.com/google/certificate-transparency/pull/766#discussion_r32306155>
> :
>
> > +  task.Wait();
> > +  EXPECT_THAT(task.status(), StatusIs(util::error::FAILED_PRECONDITION));
> > +  EXPECT_EQ(kSSLErrorStatus, resp.status_code);
> > +}
> > +
> > +
> > +TEST_F(UrlFetcherTest, TestNotListening) {
> > +  UrlFetcher::Request req(URL("https://localhost:63544"));
> > +  UrlFetcher::Response resp;
> > +
> > +  SyncTask task(&pool_);
> > +  fetcher_->Fetch(req, &resp, task.task());
> > +  task.Wait();
> > +  EXPECT_THAT(task.status(), StatusIs(util::error::FAILED_PRECONDITION));
> > +  EXPECT_EQ(kUnknownErrorStatus, resp.status_code);
> > +}
>
> Those are hard without messing with the system's DNS/hosts file.
> I did try using things likesomething.localhst.co.uk (which resolves to
> 127.0.0.1 and causes the request to hit a local openssl s_server instance
> with a * cert, but some resolvers seem to dislike that and error out rather
> than returning the address.)
>
> I'll see if I can come up with something; stub out the name resolution
> inside libevent::Base maybe.
>

Yeah. Or maybe some kind of standalone test that shows the matching works
without having to make a connection (not as good, but might be good enough,
combined with an inline test to ensure matching is called at all!).


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/certificate-transparency/pull/766/files#r32306155>
> .
>